res/controllers/Arturia-KeyLab-Mk1-scripts.js

```diff
@@ -0,0 +1,759 @@
+            // Faders
+            this.faders = new Array(9);
+            this.faders[0] = new components.Pot({
+                group: "[Channel3]",
+                midi: [0xB0, 0x49],
+                key: "volume",
+                softTakeover: softTakeover,
+            });
+            this.faders[1] = new components.Pot({
+                group: "[Channel1]",
+                midi: [0xB0, 0x4B],
+                key: "volume",
+                softTakeover: softTakeover,
+            });
+            this.faders[2] = new components.Pot({
+                group: "[Channel2]",
+                midi: [0xB0, 0x4F],
+                key: "volume",
+                softTakeover: softTakeover,
+            });
+            this.faders[3] = new components.Pot({
+                group: "[Channel4]",
+                midi: [0xB0, 0x48],
+                key: "volume",
+                softTakeover: softTakeover,
+            });
+            this.faders[4] = new components.Pot({
+                group: "[Channel3]",
+                midi: [0xB0, 0x50],
+                key: "rate",
+                softTakeover: softTakeover,
+            });
+            this.faders[5] = new components.Pot({
+                group: "[Channel1]",
+                midi: [0xB0, 0x51],
+                key: "rate",
+                softTakeover: softTakeover,
+            });
+            this.faders[6] = new components.Pot({
+                group: "[Channel2]",
+                midi: [0xB0, 0x52],
+                key: "rate",
+                softTakeover: softTakeover,
+            });
+            this.faders[7] = new components.Pot({
+                group: "[Channel4]",
+                midi: [0xB0, 0x53],
+                key: "rate",
+                softTakeover: softTakeover,
+            });
+            this.faders[8] = new components.Pot({
+                group: "[Master]",
+                midi: [0xB0, 0x55],
+                key: "crossfader",
+                softTakeover: softTakeover,
+            });
+            // Pots
+            this.mainGain = new components.Pot({
+                group: "[Master]",
+                midi: [0xB0, 0x07],
+                key: "gain",
+                softTakeover: softTakeover,
+            });
+            this.headGain = new components.Pot({
+                group: "[Master]",
+                midi: [0xB0, 0x00],
+                key: "headGain",
+                softTakeover: softTakeover,
+            });
+            // Drum pads
+            this.samplers = [];
+            for (let i = 0; i < 16; i++) {
+                this.samplers[i] = new components.SamplerButton({
+                    number: i + 1,
+                    midi: [0x99, pad(i)], // Drum pads.
+                    volumeByVelocity: true,
+                    // These numbers are made up magic constants that can be
+                    // checked in the "send" callback defined below.
```

Comment:
who made them up? Or rather, have you defined them or do they come from the firmware?

Reply:
This is all built into componentsJS. Send gets called with the value of `this.playing` or `this.empty` etc. for the purpose of setting the color. We can't set the color, so I just made up some random magic numbers and defined them here so that we'd have something to check in the switch function so we'd know what action is being sent.